tools.py

```diff
@@ -227,51 +412,156 @@ def load_journals_list(journals_file='journals.txt'):
 class XMLValidator(object):
 
     def __init__(self):
-        str_schema = open(
-            os.path.abspath(
-                os.path.join(
-                    os.path.dirname(__file__), 'xsd/ThomsonReuters_publishing.xsd')), 'r')
-        schema_doc = etree.parse(str_schema)
-        self._schema = etree.XMLSchema(schema_doc)
+        self.xsd_filename = os.path.abspath(
+                            os.path.join(
+                                os.path.dirname(__file__),
+                                'xsd/ThomsonReuters_publishing.xsd'))
+        self.articlemeta_url = 'http://articlemeta.scielo.org/api/v1/article'
+
+    def _get_xml(self, collection, code):
+        params = {'collection': collection,
+                  'code': code,
+                  'format': 'xmlwos'}
+        try:
+            return requests.get(
+                self.articlemeta_url, params=params, timeout=30).text
+        except:
+            logging.error(
+                'error fetching url: %s' % self.articlemeta_url)
 
     def validate_xml(self, collection, code):
-        articlemeta_url = 'http://articlemeta.scielo.org/api/v1/article'
+        textxml = self._get_xml(collection, code)
+        validated = ValidatedXML(textxml, self.xsd_filename)
+        article_report = ArticleReport(self.articlemeta_url, collection, code)
+        article_report.save(validated)
+        if validated.errors is None or len(validated.errors) == 0:
+            return validated.parsed
 
-        params = {'collection': collection, 'code': code, 'format': 'xmlwos'}
 
-        try:
-            textxml = requests.get(articlemeta_url, params=params, timeout=30).text
-        except:
-            logging.error('error fetching url from articlemeta: %s' % articlemeta_url)
-            return None
+class XML(object):
 
-        xml = StringIO(textxml)
+    def __init__(self, textxml):
+        self.parse_errors = []
+        self.text = textxml
+        self._parse_xml()
 
-        ## Validating well formed
+    def _parse_xml(self):
+        xml = StringIO(self.text)
+        self.parsed = None
         try:
-            parsedxml = etree.parse(xml)
-        except:
-            msg = u"%s:Not a well formed XML" % code
-            write_log(msg)
-            return None
+            self.parsed = etree.parse(xml)
+        except etree.XMLSyntaxError as e:
+            self.parse_errors.append(str(e))
+        except Exception as e:
+            self.parse_errors.append(e)
+
+    @property
+    def pretty_text(self):
+        if self.parsed is None:
+            return self.text.replace('<', '\n<').replace('\n', '\n').strip()
```

Comment:
A abordagem para casos onde ``self.parsed is None`` não atende todos os casos de XML, mas acredito que seja suficiente para os produzidos pelo formato ``xmlwos``. O último replace troca ``\n`` por ``\n``, o que não faz sentido.

Reply:
ops... obviamente deveria ser replace de outra coisa...

Reply:
```self.parsed is None``` significa que o XML não foi possível de ser carregado, entra nos casos em que o textxml is None ou o XML é inválido. Os demais casos, entra no outro return.

Trocarei ```self.parsed``` para ```self.tree```. Talvez fique mais claro.